deeplake/core/chunk/chunk_compressed_chunk.py

```diff
@@ -35,16 +36,194 @@ def __init__(self, *args, **kwargs):
+        decompressed_bytes = self.decompressed_bytes
+        while True:
+            if sample_nbytes:
+                num_samples = int(
+                    max(
+                        0,
+                        min(
+                            len(incoming_samples),
+                            (
+                                (min_chunk_size / self._compression_ratio)
+                                - len(decompressed_bytes)
+                            )
+                            // sample_nbytes,
+                        ),
+                    )
+                )
+            else:
+                num_samples = len(incoming_samples)
+            if not num_samples:
+
+                # Check if compression ratio is actually better
+                s = self._text_sample_to_byte_string(incoming_samples[0])
+                new_decompressed = decompressed_bytes + s
+                compressed_bytes = compress_bytes(
+                    new_decompressed, compression=self.compression
+                )
+
+                if len(compressed_bytes) <= min_chunk_size:
+                    self._compression_ratio /= 2
+                    continue
+                else:
+                    if self.decompressed_bytes:
+                        break
+                    else:
+                        self.decompressed_bytes = new_decompressed
+                        self._data_bytes = compressed_bytes
+                        self._changed = False
+                        num_samples = 1
+                        lengths[0] = len(s)
+                        break
+            else:
+                samples_to_chunk = incoming_samples[:num_samples]
+                bts = list(map(self._text_sample_to_byte_string, samples_to_chunk))
+                for i, b in enumerate(bts):
+                    lengths[i] = len(b)
+                self.decompressed_bytes = b"".join([decompressed_bytes, *bts])
+                del bts
+                self._changed = True
+                break
+        if num_samples:
+            lview = lengths[:num_samples]
+            csum = np.cumsum(lengths[: num_samples - 1])
+            bps = np.zeros((num_samples, 3), dtype=ENCODING_DTYPE)
+            enc = self.byte_positions_encoder
+            arr = enc._encoded
+            if len(arr):
+                last_seen = arr[-1, 2] + 1
+                if len(arr) == 1:
+                    offset = (arr[0, 2] + 1) * arr[0, 0]
+                else:
+                    offset = (arr[-1, 2] - arr[-2, 2]) * arr[-1, 0] + arr[-1, 1]
+            else:
+                last_seen = 0
+                offset = 0
+            bps[:, 2] = np.arange(last_seen, num_samples + last_seen)
+            bps[0, 1] = offset
+            bps[:, 0] = lview
+            csum += offset
+            bps[1:, 1] = csum
+            if len(arr):
+                arr = np.concatenate([arr, bps], 0)
+            else:
+                arr = bps
+            enc._encoded = arr
```

Comment:
Updating byte position encoder one sample at a time in a for loop can be very bad for large inputs where its unlikely to have samples with same byte sizes:

*  Overhead from various checks in register method of byte positions encoder.
* Numpy array concat for each new sample (as byte sizes almost always different)
*  Python for loop bad

Instead we use very efficient ops like csum, memcpy, arange etc  and do 1 final concat with the existing encoder contents.